cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -841,39 +837,56 @@ void LocalVersionedEngine::delete_trees_responsibly(
     auto data_keys_not_to_be_deleted = get_data_keys_set(store(), *not_to_delete, read_opts);
     not_to_delete.clear();
     log::version().debug("Forbidden: {} total of data keys", data_keys_not_to_be_deleted.size());
-
     RemoveOpts remove_opts;
     remove_opts.ignores_missing_key_ = true;
-
-    std::vector<entity::VariantKey> vks;
-    // Delete any associated column stats keys first
+    folly::Future<folly::Unit> write_version_fut;
     if (!dry_run) {
+        std::vector<entity::VariantKey> vks_column_stats;
+        // Delete any associated column stats keys first
         std::transform(keys_to_delete->begin(),
                        keys_to_delete->end(),
-                       std::back_inserter(vks),
+                       std::back_inserter(vks_column_stats),
                        [](const IndexTypeKey& index_key) {
             return index_key_to_column_stats_key(index_key);
         });
-        store()->remove_keys(vks, remove_opts).get();
-    }
 
-    vks.clear();
-    if (!dry_run) {
-        std::copy(keys_to_delete->begin(), keys_to_delete->end(), std::back_inserter(vks));
-        store()->remove_keys(vks, remove_opts).get();
-    }
-
-    vks.clear();
-    std::copy_if(std::make_move_iterator(data_keys_to_be_deleted.begin()),
-                 std::make_move_iterator(data_keys_to_be_deleted.end()),
-                 std::back_inserter(vks),
-                 [&](const auto& k) {return !data_keys_not_to_be_deleted.count(k);});
-    log::version().debug("Index keys deleted. Proceeding with {} number of data keys", vks.size());
-    if (!dry_run) {
-        store()->remove_keys(vks, remove_opts).get();
+        std::vector<entity::VariantKey> vks_to_delete;
+        std::copy(std::make_move_iterator(keys_to_delete->begin()),
+                    std::make_move_iterator(keys_to_delete->end()),
+                    std::back_inserter(vks_to_delete));
+
+        write_version_fut = store()->remove_keys(std::move(vks_column_stats), remove_opts)
+        .thenValue([this, vks_to_delete = std::move(vks_to_delete), remove_opts](auto&& ) mutable {
+            return store()->remove_keys(std::move(vks_to_delete), remove_opts);
+        })
+        .thenValue([](auto&&){
+            return folly::Unit();
+        });
     }
+    return std::move(write_version_fut)
```

Comment:
Returning rvalues is usually a bad idea as it prevents the compiler from doing copy elision optimizations. Especially in this case as folly::Future has its copy [constructor deleted](https://github.com/facebook/folly/blob/c409a5a381bfb88fe4aafb5270aaf24c8267cf99/folly/futures/Future.h#L1069).
```suggestion
    return write_version_fut
```

Reply:
I agree in the general idea. However, note that in this specific case you commented we were concatenating the futures (it is why the line you suggested to change does not contain ";"), which they really need to be moved when we create such a chain. For example: return std::move(A).ThenValue(...).ThenValue(...);

nevertheless, this specific part of the code has been refactored in the current version

Reply:
You have a deleted copy constructor, there is no need to move in this case. The compiler will do the right thing in this case. It is either going to move or elide.

Reply:
The line of code commented in this thread is not present anymore and as I commented it was applying to the future chain construction, which require moving. Is there any other line that concerns you? If so, please can you refer to it.